### PR TITLE
[url_launcher] Add a new launchUrl to platform interface

### PR DESCRIPTION
**Pull request by stuartmorgan**
This creates a new platform interface method for launching that closely
parallels the new public-facing API, so that implementations can switch
to implementing a more platform-neutral implementation. This will pave
the way for things like cleanly implementing
`externalNonBrowserApplication` support on non-iOS platforms.

To fully future-proof this method, it uses a wrapper parameter object
so that new types of parameters (i.e., things that aren't just new launch
modes or webview configuration) can be added to it without breaking
changes.

A follow-up will switch the app-facing package to call this new method
instead of the legacy method in the new APIs.

Implementation packages can adopt the new method as is useful for them;
eventually we can do a cleanup pass if we want to fully deprecate the
old method.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart [hunk 1]
```diff
@@ -2,69 +2,5 @@
...
-  }) {
-    throw UnimplementedError('launch() has not been implemented.');
-  }
-
-  /// Closes the WebView, if one was opened earlier by [launch].
-  Future<void> closeWebView() {
-    throw UnimplementedError('closeWebView() has not been implemented.');
-  }
-}
+export 'package:url_launcher_platform_interface/src/types.dart';
```
Comment (bparrishMines) [inline packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart:5]: nit: Should these be relative imports?

**Review by bparrishMines — APPROVED** [review]
LGTM with some nits

File: packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart]: Oops, I wonder how I did that. Fixed!

File: packages/url_launcher/url_launcher_platform_interface/lib/src/url_launcher_platform.dart [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  /// The delegate used by the Link widget to build itself.
+  LinkDelegate? get linkDelegate;
+
+  /// Returns `true` if this platform is able to launch [url].
+  Future<bool> canLaunch(String url) {
+    throw UnimplementedError('canLaunch() has not been implemented.');
+  }
+
+  /// Returns `true` if the given [url] was successfully launched.
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/url_launcher_platform.dart:52]: Added the initial sentence overview from the app-facing API.

File: packages/url_launcher/url_launcher_platform_interface/lib/src/url_launcher_platform.dart [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    required bool enableJavaScript,
+    required bool enableDomStorage,
+    required bool universalLinksOnly,
+    required Map<String, String> headers,
+    String? webOnlyWindowName,
+  }) {
+    throw UnimplementedError('launch() has not been implemented.');
+  }
+
+  /// Returns `true` if the given [url] was successfully launched.
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/url_launcher_platform.dart:71]: Same fix here.

**Comment by stuartmorgan** [comment]
> I was wondering if the new API should be `Uri`-based, but it probably doesn't add much in the way of performance or security :/

It's non-obvious here because it doesn't include the second part (maybe I should have done it as a multi-package initial PR), but it can't be; `launchUrlString` needs to use this new method too, and we can't convert to `Uri` there because that API exists for cases where `Uri` can't handle a URL that the platform can.

**Comment by ditman** [comment]
Yes, yes, I remember the `_string` API for some of the funky non-standard URLs that some users needed!

**Pull request merged**